scipy/optimize/_bglu_dense.pyx

```diff
@@ -0,0 +1,409 @@
+
+#    # Cython, high level
+#    if i == 0:
+#        H[i:i+2] = H[i+1::-1]
+#    else:
+#        H[i:i+2] = H[i+1:i-1:-1]
+
+#    # Cython, low level
+#    cdef double temp
+#    cdef int j
+#    for j in range(i, H.shape[1]):
+#        temp = H[i, j]
+#        H[i, j] = H[i+1, j]
+#        H[i+1, j] = temp
+
+    # Cython, using BLAS
+    cdef int n = H.shape[1]-i
+    cdef int one = 1
+    dswap(&n, &H[i, i], &one, &H[i+1, i], &one)
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True) # not really important
+cdef double row_subtract(self, double[:, ::1] H, int i):
+    """
+    Zeros first nonzero element of row i+1 of H by subtracting appropriate
+    multiple of row i; represents matrix product by matrix 5.10. Returns
+    factor g for storage.
+    """
+    cdef double g = H[i+1, i]/H[i,i]
+
+#        # Python
+#        H[i+1, i:] -= g*H[i, i:]
+
+#        # Cython
+#        for j in range(i, H.shape[1]):
+#            H[i+1, j] -= H[i, j]*g
+
+    # Cython, using BLAS
+    cdef int n = H.shape[1]-i
+    cdef double ng = -g
+    cdef int one = 1
+    daxpy(&n, &ng, &H[i, i], &one, &H[i+1, i], &one)
+
+    return g
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void hess_lu(self, double[:, ::1] H, int i, double[:,::1] ops):
+    """
+    Converts Hessenberg matrix H with first nonzero off-diagonal in
+    column i to upper triangular, recording elementary row operations.
+    That is, performs and records operations in Equation 5.9.
+    """
+    cdef int m = H.shape[1]
+    cdef int j, k
+    cdef double piv1, piv2
+    cdef double g
+    cdef double swap
+
+    for k in range(i, m-1):
+        j = k-i
+        piv1, piv2 = abs(H[k, k]), abs(H[k+1, k]) # np.abs(H[k:k+2,k])
+        swap = float(piv1 < piv2)
+        # swap rows to ensure |g| <= 1
+        if swap:
+            swap_rows(self, H, k)
+        g = row_subtract(self, H, k)
+
+        ops[j, 0] = swap
+        ops[j, 1] = g
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void perform_ops(self, double[::1] y, double[:,::1] ops, bint rev = False):
+    """
+    Replays operations needed to convert Hessenberg matrix into upper
+    triangular form on a vector y. Equivalent to matrix multlication by
+    inverse of matrix 5.12.
+    """
+    cdef int i, j, k, m
+    cdef double g, swap
+
+    m = y.shape[0]
+    i = m - ops.shape[0] - 1
+    if not rev:
+        for k in range(i, m-1):
+            j = k - i
+            swap = ops[j,0]
+            g = ops[j, 1]
+            if swap:
+                y[k], y[k+1] = y[k+1], y[k]
+#                swap_rows(self, y, k)
+            y[k+1] -= g*y[k]
+    else:
+        for k in range(m-2, i-1, -1):
+            j = k -i
+            swap = ops[j,0]
+            g = ops[j, 1]
+            y[k] -= g*y[k+1]
+            if swap:
+                y[k], y[k+1] = y[k+1], y[k]
+#                swap_rows(self, y, k)
+
+def _consider_refactor(method):
```

Comment:
This is a feature. It is odd, yes, but it is a feature.
The class maintains a factorization of the basis matrix which is used to solve a linear system. The initial factorization is slow, but updates to it are fast. Initially, solving the linear system using the factorization is fast, but it gets slower with each update. Ultimately we want the fastest average solve time, so the point of this is to track the average solve time. Initially, the average solve time decreases with each solve after the first. After too many updates, the average solve time begins to increase again. This refactors the matrix (starting everything from scratch) when the average solve time begins to increase again.
Too weird?

Reply:
It sounds like it would cause for non-deterministic behavior. Which is very much undesired. Could you time it once and have a fixed estimate of how many updates should be done, or is it problem-dependent?

If it's problem-dependent *and* important, there at least needs to be a way to fix the behavior (similar to e.g. setting a seed for functions using random numbers).

Reply:
It's problem-dependent. There is a way to fix the behavior - `mast=False` is the default, so by default this timing stuff is not used. It's an option if the user is willing to accept nondeterministic behavior. That's mentioned in the documentation. Is this OK?